{
  "database_name": "retryable-reads-tests",
  "collection_name": "coll",
  "data": [
    {
      "_id": 1,
      "x": 0
    },
    {
      "_id": 2,
      "x": 1
    },
    {
      "_id": 3,
      "x": 2
    }
  ],
  "minServerVersion": "4.1.7",
  "tests": [
    {
      "description": "MapReduce succeeds after InterruptedAtShutdown",
      "clientOptions": {
        "retryReads": true
      },
      "failPoint": {
        "configureFailPoint": "failCommand",
        "mode": {
          "times": 1
        },
        "data": {
          "failCommands": [
            "mapReduce"
          ],
          "errorCode": 11600
        }
      },
      "operations": [
        {
          "name": "mapReduce",
          "object": "collection",
          "arguments": {
            "map": {
              "$code": "function inc() { return emit(0, this.x + 1) }"
            },
            "reduce": {
              "$code": "function sum(key, values) { return values.reduce((acc, x) => acc + x); }"
            },
            "out": {
              "inline": 1
            }
          },
          "result": [
            {
              "_id": 0,
              "value": 6
            }
          ]
        }
      ],
      "expectations": [
        {
          "command_started_event": {
            "command": {
              "mapReduce": "coll",
              "map": {
                "$code": "function inc() { return emit(0, this.x + 1) }"
              },
              "reduce": {
                "$code": "function sum(key, values) { return values.reduce((acc, x) => acc + x); }"
              },
              "out": {
                "inline": 1
              }
            },
            "database_name": "retryable-reads-tests"
          }
        },
        {
          "command_started_event": {
            "command": {
              "mapReduce": "coll",
              "map": {
                "$code": "function inc() { return emit(0, this.x + 1) }"
              },
              "reduce": {
                "$code": "function sum(key, values) { return values.reduce((acc, x) => acc + x); }"
              },
              "out": {
                "inline": 1
              }
            },
            "database_name": "retryable-reads-tests"
          }
        }
      ]
    },
    {
      "description": "MapReduce succeeds after InterruptedDueToStepDown",
      "clientOptions": {
        "retryReads": true
      },
      "failPoint": {
        "configureFailPoint": "failCommand",
        "mode": {
          "times": 1
        },
        "data": {
          "failCommands": [
            "mapReduce"
          ],
          "errorCode": 11602
        }
      },
      "operations": [
        {
          "name": "mapReduce",
          "object": "collection",
          "arguments": {
            "map": {
              "$code": "function inc() { return emit(0, this.x + 1) }"
            },
            "reduce": {
              "$code": "function sum(key, values) { return values.reduce((acc, x) => acc + x); }"
            },
            "out": {
              "inline": 1
            }
          },
          "result": [
            {
              "_id": 0,
              "value": 6
            }
          ]
        }
      ],
      "expectations": [
        {
          "command_started_event": {
            "command": {
              "mapReduce": "coll",
              "map": {
                "$code": "function inc() { return emit(0, this.x + 1) }"
              },
              "reduce": {
                "$code": "function sum(key, values) { return values.reduce((acc, x) => acc + x); }"
              },
              "out": {
                "inline": 1
              }
            },
            "database_name": "retryable-reads-tests"
          }
        },
        {
          "command_started_event": {
            "command": {
              "mapReduce": "coll",
              "map": {
                "$code": "function inc() { return emit(0, this.x + 1) }"
              },
              "reduce": {
                "$code": "function sum(key, values) { return values.reduce((acc, x) => acc + x); }"
              },
              "out": {
                "inline": 1
              }
            },
            "database_name": "retryable-reads-tests"
          }
        }
      ]
    },
    {
      "description": "MapReduce succeeds after NotMaster",
      "clientOptions": {
        "retryReads": true
      },
      "failPoint": {
        "configureFailPoint": "failCommand",
        "mode": {
          "times": 1
        },
        "data": {
          "failCommands": [
            "mapReduce"
          ],
          "errorCode": 10107
        }
      },
      "operations": [
        {
          "name": "mapReduce",
          "object": "collection",
          "arguments": {
            "map": {
              "$code": "function inc() { return emit(0, this.x + 1) }"
            },
            "reduce": {
              "$code": "function sum(key, values) { return values.reduce((acc, x) => acc + x); }"
            },
            "out": {
              "inline": 1
            }
          },
          "result": [
            {
              "_id": 0,
              "value": 6
            }
          ]
        }
      ],
      "expectations": [
        {
          "command_started_event": {
            "command": {
              "mapReduce": "coll",
              "map": {
                "$code": "function inc() { return emit(0, this.x + 1) }"
              },
              "reduce": {
                "$code": "function sum(key, values) { return values.reduce((acc, x) => acc + x); }"
              },
              "out": {
                "inline": 1
              }
            },
            "database_name": "retryable-reads-tests"
          }
        },
        {
          "command_started_event": {
            "command": {
              "mapReduce": "coll",
              "map": {
                "$code": "function inc() { return emit(0, this.x + 1) }"
              },
              "reduce": {
                "$code": "function sum(key, values) { return values.reduce((acc, x) => acc + x); }"
              },
              "out": {
                "inline": 1
              }
            },
            "database_name": "retryable-reads-tests"
          }
        }
      ]
    },
    {
      "description": "MapReduce succeeds after NotMasterNoSlaveOk",
      "clientOptions": {
        "retryReads": true
      },
      "failPoint": {
        "configureFailPoint": "failCommand",
        "mode": {
          "times": 1
        },
        "data": {
          "failCommands": [
            "mapReduce"
          ],
          "errorCode": 13435
        }
      },
      "operations": [
        {
          "name": "mapReduce",
          "object": "collection",
          "arguments": {
            "map": {
              "$code": "function inc() { return emit(0, this.x + 1) }"
            },
            "reduce": {
              "$code": "function sum(key, values) { return values.reduce((acc, x) => acc + x); }"
            },
            "out": {
              "inline": 1
            }
          },
          "result": [
            {
              "_id": 0,
              "value": 6
            }
          ]
        }
      ],
      "expectations": [
        {
          "command_started_event": {
            "command": {
              "mapReduce": "coll",
              "map": {
                "$code": "function inc() { return emit(0, this.x + 1) }"
              },
              "reduce": {
                "$code": "function sum(key, values) { return values.reduce((acc, x) => acc + x); }"
              },
              "out": {
                "inline": 1
              }
            },
            "database_name": "retryable-reads-tests"
          }
        },
        {
          "command_started_event": {
            "command": {
              "mapReduce": "coll",
              "map": {
                "$code": "function inc() { return emit(0, this.x + 1) }"
              },
              "reduce": {
                "$code": "function sum(key, values) { return values.reduce((acc, x) => acc + x); }"
              },
              "out": {
                "inline": 1
              }
            },
            "database_name": "retryable-reads-tests"
          }
        }
      ]
    },
    {
      "description": "MapReduce succeeds after NotMasterOrSecondary",
      "clientOptions": {
        "retryReads": true
      },
      "failPoint": {
        "configureFailPoint": "failCommand",
        "mode": {
          "times": 1
        },
        "data": {
          "failCommands": [
            "mapReduce"
          ],
          "errorCode": 13436
        }
      },
      "operations": [
        {
          "name": "mapReduce",
          "object": "collection",
          "arguments": {
            "map": {
              "$code": "function inc() { return emit(0, this.x + 1) }"
            },
            "reduce": {
              "$code": "function sum(key, values) { return values.reduce((acc, x) => acc + x); }"
            },
            "out": {
              "inline": 1
            }
          },
          "result": [
            {
              "_id": 0,
              "value": 6
            }
          ]
        }
      ],
      "expectations": [
        {
          "command_started_event": {
            "command": {
              "mapReduce": "coll",
              "map": {
                "$code": "function inc() { return emit(0, this.x + 1) }"
              },
              "reduce": {
                "$code": "function sum(key, values) { return values.reduce((acc, x) => acc + x); }"
              },
              "out": {
                "inline": 1
              }
            },
            "database_name": "retryable-reads-tests"
          }
        },
        {
          "command_started_event": {
            "command": {
              "mapReduce": "coll",
              "map": {
                "$code": "function inc() { return emit(0, this.x + 1) }"
              },
              "reduce": {
                "$code": "function sum(key, values) { return values.reduce((acc, x) => acc + x); }"
              },
              "out": {
                "inline": 1
              }
            },
            "database_name": "retryable-reads-tests"
          }
        }
      ]
    },
    {
      "description": "MapReduce succeeds after PrimarySteppedDown",
      "clientOptions": {
        "retryReads": true
      },
      "failPoint": {
        "configureFailPoint": "failCommand",
        "mode": {
          "times": 1
        },
        "data": {
          "failCommands": [
            "mapReduce"
          ],
          "errorCode": 189
        }
      },
      "operations": [
        {
          "name": "mapReduce",
          "object": "collection",
          "arguments": {
            "map": {
              "$code": "function inc() { return emit(0, this.x + 1) }"
            },
            "reduce": {
              "$code": "function sum(key, values) { return values.reduce((acc, x) => acc + x); }"
            },
            "out": {
              "inline": 1
            }
          },
          "result": [
            {
              "_id": 0,
              "value": 6
            }
          ]
        }
      ],
      "expectations": [
        {
          "command_started_event": {
            "command": {
              "mapReduce": "coll",
              "map": {
                "$code": "function inc() { return emit(0, this.x + 1) }"
              },
              "reduce": {
                "$code": "function sum(key, values) { return values.reduce((acc, x) => acc + x); }"
              },
              "out": {
                "inline": 1
              }
            },
            "database_name": "retryable-reads-tests"
          }
        },
        {
          "command_started_event": {
            "command": {
              "mapReduce": "coll",
              "map": {
                "$code": "function inc() { return emit(0, this.x + 1) }"
              },
              "reduce": {
                "$code": "function sum(key, values) { return values.reduce((acc, x) => acc + x); }"
              },
              "out": {
                "inline": 1
              }
            },
            "database_name": "retryable-reads-tests"
          }
        }
      ]
    },
    {
      "description": "MapReduce succeeds after ShutdownInProgress",
      "clientOptions": {
        "retryReads": true
      },
      "failPoint": {
        "configureFailPoint": "failCommand",
        "mode": {
          "times": 1
        },
        "data": {
          "failCommands": [
            "mapReduce"
          ],
          "errorCode": 91
        }
      },
      "operations": [
        {
          "name": "mapReduce",
          "object": "collection",
          "arguments": {
            "map": {
              "$code": "function inc() { return emit(0, this.x + 1) }"
            },
            "reduce": {
              "$code": "function sum(key, values) { return values.reduce((acc, x) => acc + x); }"
            },
            "out": {
              "inline": 1
            }
          },
          "result": [
            {
              "_id": 0,
              "value": 6
            }
          ]
        }
      ],
      "expectations": [
        {
          "command_started_event": {
            "command": {
              "mapReduce": "coll",
              "map": {
                "$code": "function inc() { return emit(0, this.x + 1) }"
              },
              "reduce": {
                "$code": "function sum(key, values) { return values.reduce((acc, x) => acc + x); }"
              },
              "out": {
                "inline": 1
              }
            },
            "database_name": "retryable-reads-tests"
          }
        },
        {
          "command_started_event": {
            "command": {
              "mapReduce": "coll",
              "map": {
                "$code": "function inc() { return emit(0, this.x + 1) }"
              },
              "reduce": {
                "$code": "function sum(key, values) { return values.reduce((acc, x) => acc + x); }"
              },
              "out": {
                "inline": 1
              }
            },
            "database_name": "retryable-reads-tests"
          }
        }
      ]
    },
    {
      "description": "MapReduce succeeds after HostNotFound",
      "clientOptions": {
        "retryReads": true
      },
      "failPoint": {
        "configureFailPoint": "failCommand",
        "mode": {
          "times": 1
        },
        "data": {
          "failCommands": [
            "mapReduce"
          ],
          "errorCode": 7
        }
      },
      "operations": [
        {
          "name": "mapReduce",
          "object": "collection",
          "arguments": {
            "map": {
              "$code": "function inc() { return emit(0, this.x + 1) }"
            },
            "reduce": {
              "$code": "function sum(key, values) { return values.reduce((acc, x) => acc + x); }"
            },
            "out": {
              "inline": 1
            }
          },
          "result": [
            {
              "_id": 0,
              "value": 6
            }
          ]
        }
      ],
      "expectations": [
        {
          "command_started_event": {
            "command": {
              "mapReduce": "coll",
              "map": {
                "$code": "function inc() { return emit(0, this.x + 1) }"
              },
              "reduce": {
                "$code": "function sum(key, values) { return values.reduce((acc, x) => acc + x); }"
              },
              "out": {
                "inline": 1
              }
            },
            "database_name": "retryable-reads-tests"
          }
        },
        {
          "command_started_event": {
            "command": {
              "mapReduce": "coll",
              "map": {
                "$code": "function inc() { return emit(0, this.x + 1) }"
              },
              "reduce": {
                "$code": "function sum(key, values) { return values.reduce((acc, x) => acc + x); }"
              },
              "out": {
                "inline": 1
              }
            },
            "database_name": "retryable-reads-tests"
          }
        }
      ]
    },
    {
      "description": "MapReduce succeeds after HostUnreachable",
      "clientOptions": {
        "retryReads": true
      },
      "failPoint": {
        "configureFailPoint": "failCommand",
        "mode": {
          "times": 1
        },
        "data": {
          "failCommands": [
            "mapReduce"
          ],
          "errorCode": 6
        }
      },
      "operations": [
        {
          "name": "mapReduce",
          "object": "collection",
          "arguments": {
            "map": {
              "$code": "function inc() { return emit(0, this.x + 1) }"
            },
            "reduce": {
              "$code": "function sum(key, values) { return values.reduce((acc, x) => acc + x); }"
            },
            "out": {
              "inline": 1
            }
          },
          "result": [
            {
              "_id": 0,
              "value": 6
            }
          ]
        }
      ],
      "expectations": [
        {
          "command_started_event": {
            "command": {
              "mapReduce": "coll",
              "map": {
                "$code": "function inc() { return emit(0, this.x + 1) }"
              },
              "reduce": {
                "$code": "function sum(key, values) { return values.reduce((acc, x) => acc + x); }"
              },
              "out": {
                "inline": 1
              }
            },
            "database_name": "retryable-reads-tests"
          }
        },
        {
          "command_started_event": {
            "command": {
              "mapReduce": "coll",
              "map": {
                "$code": "function inc() { return emit(0, this.x + 1) }"
              },
              "reduce": {
                "$code": "function sum(key, values) { return values.reduce((acc, x) => acc + x); }"
              },
              "out": {
                "inline": 1
              }
            },
            "database_name": "retryable-reads-tests"
          }
        }
      ]
    },
    {
      "description": "MapReduce succeeds after NetworkTimeout",
      "clientOptions": {
        "retryReads": true
      },
      "failPoint": {
        "configureFailPoint": "failCommand",
        "mode": {
          "times": 1
        },
        "data": {
          "failCommands": [
            "mapReduce"
          ],
          "errorCode": 89
        }
      },
      "operations": [
        {
          "name": "mapReduce",
          "object": "collection",
          "arguments": {
            "map": {
              "$code": "function inc() { return emit(0, this.x + 1) }"
            },
            "reduce": {
              "$code": "function sum(key, values) { return values.reduce((acc, x) => acc + x); }"
            },
            "out": {
              "inline": 1
            }
          },
          "result": [
            {
              "_id": 0,
              "value": 6
            }
          ]
        }
      ],
      "expectations": [
        {
          "command_started_event": {
            "command": {
              "mapReduce": "coll",
              "map": {
                "$code": "function inc() { return emit(0, this.x + 1) }"
              },
              "reduce": {
                "$code": "function sum(key, values) { return values.reduce((acc, x) => acc + x); }"
              },
              "out": {
                "inline": 1
              }
            },
            "database_name": "retryable-reads-tests"
          }
        },
        {
          "command_started_event": {
            "command": {
              "mapReduce": "coll",
              "map": {
                "$code": "function inc() { return emit(0, this.x + 1) }"
              },
              "reduce": {
                "$code": "function sum(key, values) { return values.reduce((acc, x) => acc + x); }"
              },
              "out": {
                "inline": 1
              }
            },
            "database_name": "retryable-reads-tests"
          }
        }
      ]
    },
    {
      "description": "MapReduce succeeds after SocketException",
      "clientOptions": {
        "retryReads": true
      },
      "failPoint": {
        "configureFailPoint": "failCommand",
        "mode": {
          "times": 1
        },
        "data": {
          "failCommands": [
            "mapReduce"
          ],
          "errorCode": 9001
        }
      },
      "operations": [
        {
          "name": "mapReduce",
          "object": "collection",
          "arguments": {
            "map": {
              "$code": "function inc() { return emit(0, this.x + 1) }"
            },
            "reduce": {
              "$code": "function sum(key, values) { return values.reduce((acc, x) => acc + x); }"
            },
            "out": {
              "inline": 1
            }
          },
          "result": [
            {
              "_id": 0,
              "value": 6
            }
          ]
        }
      ],
      "expectations": [
        {
          "command_started_event": {
            "command": {
              "mapReduce": "coll",
              "map": {
                "$code": "function inc() { return emit(0, this.x + 1) }"
              },
              "reduce": {
                "$code": "function sum(key, values) { return values.reduce((acc, x) => acc + x); }"
              },
              "out": {
                "inline": 1
              }
            },
            "database_name": "retryable-reads-tests"
          }
        },
        {
          "command_started_event": {
            "command": {
              "mapReduce": "coll",
              "map": {
                "$code": "function inc() { return emit(0, this.x + 1) }"
              },
              "reduce": {
                "$code": "function sum(key, values) { return values.reduce((acc, x) => acc + x); }"
              },
              "out": {
                "inline": 1
              }
            },
            "database_name": "retryable-reads-tests"
          }
        }
      ]
    },
    {
      "description": "MapReduce fails after two NotMaster errors",
      "clientOptions": {
        "retryReads": true
      },
      "failPoint": {
        "configureFailPoint": "failCommand",
        "mode": {
          "times": 2
        },
        "data": {
          "failCommands": [
            "mapReduce"
          ],
          "errorCode": 10107
        }
      },
      "operations": [
        {
          "name": "mapReduce",
          "object": "collection",
          "arguments": {
            "map": {
              "$code": "function inc() { return emit(0, this.x + 1) }"
            },
            "reduce": {
              "$code": "function sum(key, values) { return values.reduce((acc, x) => acc + x); }"
            },
            "out": {
              "inline": 1
            }
          },
          "error": true
        }
      ],
      "expectations": [
        {
          "command_started_event": {
            "command": {
              "mapReduce": "coll",
              "map": {
                "$code": "function inc() { return emit(0, this.x + 1) }"
              },
              "reduce": {
                "$code": "function sum(key, values) { return values.reduce((acc, x) => acc + x); }"
              },
              "out": {
                "inline": 1
              }
            },
            "database_name": "retryable-reads-tests"
          }
        },
        {
          "command_started_event": {
            "command": {
              "mapReduce": "coll",
              "map": {
                "$code": "function inc() { return emit(0, this.x + 1) }"
              },
              "reduce": {
                "$code": "function sum(key, values) { return values.reduce((acc, x) => acc + x); }"
              },
              "out": {
                "inline": 1
              }
            },
            "database_name": "retryable-reads-tests"
          }
        }
      ]
    },
    {
      "description": "MapReduce fails after NotMaster when retryReads is false",
      "clientOptions": {
        "retryReads": false
      },
      "failPoint": {
        "configureFailPoint": "failCommand",
        "mode": {
          "times": 1
        },
        "data": {
          "failCommands": [
            "mapReduce"
          ],
          "errorCode": 10107
        }
      },
      "operations": [
        {
          "name": "mapReduce",
          "object": "collection",
          "arguments": {
            "map": {
              "$code": "function inc() { return emit(0, this.x + 1) }"
            },
            "reduce": {
              "$code": "function sum(key, values) { return values.reduce((acc, x) => acc + x); }"
            },
            "out": {
              "inline": 1
            }
          },
          "error": true
        }
      ],
      "expectations": [
        {
          "command_started_event": {
            "command": {
              "mapReduce": "coll",
              "map": {
                "$code": "function inc() { return emit(0, this.x + 1) }"
              },
              "reduce": {
                "$code": "function sum(key, values) { return values.reduce((acc, x) => acc + x); }"
              },
              "out": {
                "inline": 1
              }
            },
            "database_name": "retryable-reads-tests"
          }
        }
      ]
    }
  ]
}
